.github/workflows/build_and_test.yml

```diff
@@ -18,58 +20,86 @@ jobs:
 
       - name: Run end-to-end tests
         run: JDK_9="$JAVA_HOME" python/e2e-tests/run.sh
-        if: always()
+        if: ${{ matrix.testTask == 'pythonTest' }}  # TODO: run e2e tests for MicroPython too (#85)
 
       - name: Run box tests (succeed even if they fail)
         run: JDK_9="$JAVA_HOME" ./gradlew :python:box.tests:${{ matrix.testTask }} || true
-        if: always()
       - name: Generate box tests reports
-        run: python/experiments/generate-box-tests-reports.main.kts --test-task=${{ matrix.testTask }} --failed-tests-report-path=failed-tests.txt --box-tests-report-path=box-tests-report.tsv --failure-count-report-path=failure-count.tsv --git-history-plot-path=git-history-plot.svg
-        if: always()
-      - name: Compare failed-tests.txt
-        run: diff -u failed-tests.txt python/box.tests/reports/${{ matrix.testTask }}/failed-tests.txt
-        if: always()
-      - name: Compare box-tests-report.tsv
-        run: diff -u box-tests-report.tsv python/box.tests/reports/${{ matrix.testTask }}/box-tests-report.tsv
-        if: always()
-      - name: Compare failure-count.tsv
-        run: diff -u failure-count.tsv python/box.tests/reports/${{ matrix.testTask }}/failure-count.tsv
-        if: always()
+        run: python/experiments/generate-box-tests-reports.main.kts --test-task=${{ matrix.testTask }}
+
+      - name: Upload common artifact
+        if: ${{ matrix.testTask == 'pythonTest' }}
+        uses: actions/upload-artifact@v2
+        with:
+          name: common-artifact
+          path: |
+            python/box.tests/reports/git-history-plot.svg
+            python/experiments/missing.txt
+            python/experiments/out-ir.js
+            python/experiments/out_ir.py
+      - name: Upload ${{ matrix.testTask }} artifact
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ matrix.testTask }}-artifact
+          path: python/box.tests/reports/${{ matrix.testTask }}
+
+  update_reports:
+    needs: build_and_test
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out
+        uses: actions/checkout@v2
+
+      - name: Download common artifact
+        uses: actions/download-artifact@v2
+        with:
+          name: common-artifact
+          path: python
+      - name: Download pythonTest artifact
+        uses: actions/download-artifact@v2
+        with:
+          name: pythonTest-artifact
+          path: python/box.tests/reports/pythonTest
+      - name: Download microPythonTest artifact
+        uses: actions/download-artifact@v2
+        with:
+          name: microPythonTest-artifact
+          path: python/box.tests/reports/microPythonTest
+
+      - name: Commit and push
+        run: |
+          git config --global user.email "<>"
+          git config --global user.name "GitHub Actions Bot"
+          git add python/box.tests/reports python/experiments
+          git commit --allow-empty -m "Update reports for $GITHUB_SHA"
```

Comment:
Regarding `--allow-empty`: it would be nice to produce no commit at all if no changes in the reports occur. It will decrease the noise in git history.

Reply:
Well, FMPOV, an empty commit ensures that reports are up-to-date, so it's not a noise. If we want less such auto-generated commits, we can push multiple commits at once, for example. So let me ask one more time: do you insist that we should avoid creating empty commits?

Reply:
Ok, got your PoV now - the empty commit actually brings some value. I'm fine with leaving it as is, optionally with a comment here to summarize what we concluded.